src/sagemaker_training/torch_distributed.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2018-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License'). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains functionality related to Torch Distributed Elastic Runner.
+Refer: https://pytorch.org/docs/stable/elastic/run.html
+"""
+from __future__ import absolute_import
+
+import os
+import sys
+
+from sagemaker_training import _entry_point_type, environment, errors, logging_config, process
+
+TORCH_DISTRIBUTED_MODULE = "torchrun"
+MASTER_PORT = "7777"
+RDZV_BACKEND = "c10d"
+RDZV_ID = "PT_TORCHRUN"
+
+logger = logging_config.get_logger()
+
+
+def python_executable():
+    """Return the real path for the Python executable, if it exists.
+    Return RuntimeError otherwise.
+
+    Returns:
+        (str): The real path of the current Python executable.
+    """
+    if not sys.executable:
+        raise RuntimeError("Failed to retrieve the real path for the Python executable")
+    return sys.executable
+
+
+class TorchDistributedRunner(process.ProcessRunner):
+    """Runner responsible for preparing Pytorch distributed data parallel training"""
+
+    def __init__(
+        self,
+        user_entry_point,
+        args,
+        env_vars,
+        processes_per_host,
+        master_hostname,
+        hosts,
+        current_host,
+        network_interface_name,
+        instance_type="ml.trn1.2xlarge",
+    ):
+        """Initialize a Native PT Launcher, which is responsible for executing
+        the user entry point within a process.
+
+        Args:
+            user_entry_point (str): The name of the user entry point.
+            args ([str]): A list of arguments to include when executing the entry point.
+            env_vars (dict(str,str)): A dictionary of environment variables.
+        """
+        super(TorchDistributedRunner, self).__init__(
+            user_entry_point, args, env_vars, processes_per_host
+        )
+
+        self._master_hostname = master_hostname
+        self._hosts = hosts
+        self._current_host = current_host
+        self._network_interface_name = network_interface_name
+        self._rdzv_backend = RDZV_BACKEND
+        self._instance_type = instance_type
+
+    def _setup(self):
+        logger.info("Starting distributed training through torchrun")
```

Comment:
Can we set the EFA stuff only for EFA enabled instances 
if instance_type in SM_EFA_NCCL_INSTANCES

Reply:
Since torchrun is enabled only for trainium instances, AWS Neuron documentation examples show setting of EFA related environment variables. Am I missing something here?

Reply:
After discussing offline, enabled the EFA settings only on 32xlarge